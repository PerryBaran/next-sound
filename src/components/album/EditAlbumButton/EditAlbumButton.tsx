"use client"

import Link from "next/link"
import { useUserContext } from "../../../context/UserContext"

interface Props {
  albumId: string
  albumUserId: string
  profile: boolean
}

export default function EditAlbumButton({
  albumId,
  albumUserId,
  profile
}: Props) {
  const {
    user: { id }
  } = useUserContext()

  if (!profile || albumUserId !== id) return null

  return <Link href={`/edit/album/${albumId}`}>Edit Album</Link>
}
