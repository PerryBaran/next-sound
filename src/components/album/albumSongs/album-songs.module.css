.container {
  list-style: none;
  display: flex;
  justify-content: space-between;
  margin-top: 5px;
  border-bottom: 1px solid var(--dark-grey);
}

.container:last-of-type {
  border-bottom: none;
}
