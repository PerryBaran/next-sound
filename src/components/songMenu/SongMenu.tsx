"use client"

import { useState } from "react"
import css from "./songMenu.module.css"
import { usePlaylistContext } from "../../context/PlaylistContext"
import { play, addList, dots } from "../../media/icons/index"
import Image from "next/image"

interface Props {
  songs: {
    songName: string
    audio: string
    image?: string
    artistName: string
    albumName: string
  }[]
}

export default function SongMenu({ songs }: Props) {
  const { handleAddToPlaylist } = usePlaylistContext()
  const [visible, setVisible] = useState(false)

  const handleBlur = () => {
    setVisible(false)
  }

  return (
    <div className={css.container} onBlur={handleBlur}>
      {visible ? (
        <div className={css.options}>
          <button onMouseDown={() => handleAddToPlaylist(songs, true, true)}>
            <span className={css["song-menu-image-container"]}>
              <Image src={play} alt="play now" width={9} />
            </span>
            <span>Play Now</span>
          </button>
          <button
            onMouseDown={() => handleAddToPlaylist(songs, true, false)}
            className={css["play-next"]}
          >
            <span className={css["song-menu-image-container"]}>
              <Image src={addList} alt="play now" width={15} />
            </span>
            <span>Play Next</span>
          </button>
          <button onMouseDown={() => handleAddToPlaylist(songs, false, false)}>
            <span className={css["song-menu-image-container"]}>
              <Image src={addList} alt="play now" width={15} />
            </span>
            <span>Add To Queue</span>
          </button>
        </div>
      ) : null}
      <button
        onClick={() => setVisible((prev) => !prev)}
        className={css.button}
      >
        <Image src={dots} alt="..." height={18} width={18} />
      </button>
    </div>
  )
}
