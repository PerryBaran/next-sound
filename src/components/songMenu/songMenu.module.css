.container {
  position: relative;
}

.options {
  position: absolute;
  bottom: 20px;
  right: 0px;
  width: 110px;
  display: flex;
  flex-direction: column;
  border: 1px solid var(--darker-grey);
  overflow: hidden;
  z-index: 1000;
}

.options > button {
  border: none;
  border-bottom: 1px solid var(--darker-grey);
  background-color: var(--almost-white);
  color: var(--darker-grey);
  font-weight: 100;
  cursor: pointer;
  text-align: left;
  display: flex;
  height: 20px;
}

.options > button:hover {
  background-color: var(--main-colour);
  color: var(--almost-white);
}

.options > button:last-child {
  border-bottom: none;
}

.image-container {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20px;
  height: 100%;
}

.options > button:hover img {
  filter: invert();
}

.play-next img {
  transform: scaleY(-1);
}

.button {
  border: none;
  background: none;
  padding: 5px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
}

.button > img {
  transition-duration: 250ms;
  transform: rotate(90deg);
}

.button:hover > img {
  transform: rotate(90deg) scale(1.25);
}
