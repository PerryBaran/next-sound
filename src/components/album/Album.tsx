import Image from 'next/image'
import css from './album.module.css'
import { music } from '@/media/icons'
import Link from 'next/link'
import Songs from './songs/Songs'
import SongMenu from '../songMenu/SongMenu'

interface Song {
  name: string
  url: string
  id: string
}

interface Props {
  artistName: string
  albumName: string
  albumArt?: string
  songs: Song[]
  albumUserId: string
  albumId: string
}

export default function Album({
  artistName,
  albumName,
  albumArt,
  songs,
  albumUserId,
  albumId
}: Props) {
  const album = songs.map((song) => {
    return {
      songName: song.name,
      audio: song.url,
      image: albumArt,
      artistName,
      albumName
    }
  })

  return (
    <div className={css.container}>
      <div className={css['album-info-container']}>
        <Image
          src={albumArt || music}
          alt={`${albumName} cover art`}
          height={100}
          width={100}
        />
        <div className={css['album-info']}>
          <h2>
            <Link href={`/profile/${artistName}`}>{artistName}</Link>
          </h2>
          <h3>{albumName}</h3>
        </div>
        <SongMenu songs={album} />
      </div>
      <ul className={css['songs-container']}>
        {songs.map((song) => {
          return (
            <Songs
              key={song.id}
              songName={song.name}
              audio={song.url}
              image={albumArt}
              artistName={artistName}
              albumName={albumName}
            />
          )
        })}
      </ul>
    </div>
  )
}
