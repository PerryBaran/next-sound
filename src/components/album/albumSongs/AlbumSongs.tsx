import css from "./album-songs.module.css"
import SongMenu from "../../songMenu/SongMenu"

interface Props {
  songName: string
  audio: string
  image?: string
  artistName: string
  albumName: string
}

export default function AlbumSongs(props: Props) {
  return (
    <li className={css["container"]}>
      <h4>{props.songName}</h4>
      <SongMenu songs={[props]} />
    </li>
  )
}
