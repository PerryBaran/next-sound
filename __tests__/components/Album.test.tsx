import { render, screen } from "@testing-library/react"
import Album from "../../src/components/album/Album"
import * as AlbumSongs from "../../src/components/album/albumSongs/AlbumSongs"
import * as SongMenu from "../../src/components/songMenu/SongMenu"
import * as EditAlbumButton from "../../src/components/album/EditAlbumButton/EditAlbumButton"

describe("Album", () => {
  const props = {
    artistName: "artist-name",
    albumName: "album-name",
    albumArt: "/album-art",
    songs: [
      {
        name: "song-name-1",
        url: "audio-1",
        id: "song-id-1"
      },
      {
        name: "song-name-2",
        url: "audio-2",
        id: "song-id-2"
      }
    ],
    albumUserId: "album-user-id",
    albumId: "album-id",
    profile: true
  }

  let mockedAlbumSongs = jest.fn()
  let mockedSongMenu = jest.fn()
  let mockedEditAlbumButton = jest.fn()

  jest.spyOn(SongMenu, "default").mockImplementation((props: any) => {
    mockedSongMenu(props)
    return <div />
  })

  jest.spyOn(AlbumSongs, "default").mockImplementation((props: any) => {
    mockedAlbumSongs(props)
    return <div />
  })

  jest.spyOn(EditAlbumButton, "default").mockImplementation((props: any) => {
    mockedEditAlbumButton(props)
    return <div />
  })

  test("if songs array length is 0 and profile is false, returns null", () => {
    const { container } = render(
      <Album
        artistName={props.artistName}
        albumName={props.albumName}
        albumArt={props.albumArt}
        songs={[]}
        albumUserId={props.albumUserId}
        albumId={props.albumId}
        profile={false}
      />
    )

    expect(container).toBeEmptyDOMElement()
  })

  test("renders correctly", () => {
    render(<Album {...props} />)

    expect(screen.getByAltText(`${props.albumName} cover art`)).toHaveAttribute(
      "src",
      `/_next/image?url=%2F${props.albumArt.substring(
        1,
        props.albumArt.length
      )}&w=256&q=75`
    )
    expect(screen.getByText(props.artistName)).toHaveAttribute(
      "href",
      `/profile/${props.artistName}`
    )
    expect(screen.getByText(props.albumName)).toBeTruthy()
    expect(screen.getByRole("list")).toBeTruthy()

    expect(mockedSongMenu).toBeCalledWith({ songs: props.songs.map((song) => {
      return {
        songName: song.name,
        audio: song.url,
        image: props.albumArt,
        artistName: props.artistName,
        albumName: props.albumName
      }
    })})
    expect(mockedEditAlbumButton).toBeCalledWith({
      albumId: props.albumId,
      albumUserId: props.albumUserId,
      profile: props.profile
    })
    expect(mockedAlbumSongs).toBeCalledTimes(props.songs.length)
    props.songs.forEach((song) => {
      expect(mockedAlbumSongs).toHaveBeenCalledWith({
        songName: song.name,
        audio: song.url,
        image: props.albumArt,
        artistName: props.artistName,
        albumName: props.albumName
      })
    })
  })

  describe("snapshots", () => {
    test("if songs array length is 0 and profile is false", () => {
      const { asFragment } = render(
        <Album
          artistName={props.artistName}
          albumName={props.albumName}
          albumArt={props.albumArt}
          songs={[]}
          albumUserId={props.albumUserId}
          albumId={props.albumId}
          profile={false}
        />
      )

      expect(asFragment()).toMatchSnapshot()
    })

    test("if songs array length is greater than 0 and profile is truthy", () => {
      const { asFragment } = render(<Album {...props} />)

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
