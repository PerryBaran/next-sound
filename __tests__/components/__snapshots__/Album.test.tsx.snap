// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album snapshots if songs array length is 0 and profile is false 1`] = `<DocumentFragment />`;

exports[`Album snapshots if songs array length is greater than 0 and profile is truthy 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="album-info-container"
    >
      <img
        alt="album-name cover art"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2Falbum-art&w=256&q=75"
        srcset="/_next/image?url=%2Falbum-art&w=128&q=75 1x, /_next/image?url=%2Falbum-art&w=256&q=75 2x"
        style="color: transparent;"
        width="100"
      />
      <div
        class="album-info"
      >
        <h2>
          <a
            href="/profile/artist-name"
          >
            artist-name
          </a>
        </h2>
        <h3>
          album-name
        </h3>
      </div>
      <div />
      <div />
    </div>
    <ul
      class="songs-container"
    >
      <div />
      <div />
    </ul>
  </div>
</DocumentFragment>
`;
